bazelisk.py

```diff
@@ -116,25 +122,160 @@ def normalized_machine_arch_name():
   return machine
 
 
-def determine_url(version, bazel_filename):
+SubprocessResult = collections.namedtuple("SubprocessResult", ("exit_code",))
+
+
+def subprocess_run(command, input=None, error_message=None):
+  """Kind of like Python 3's subprocess.run, but works in Python 2.
+
+  The contents of stdout and stderr are captured. If the command
+  succeeds (exit code 0), they are not printed. If the command fails,
+  stderr is printed along with the provided error message (if any).
+
+  Args:
+    command: The command to be executed, as a list of strings
+    input: A bytestring to use as stdin, or None.
+    error_message: If not None, will be logged on failure.
+
+  Returns:
+    A `SubprocessResult` including the process's exit code.
+  """
+  process = subprocess.Popen(
+      command,
+      stdin=subprocess.PIPE,
+      stdout=subprocess.PIPE,
+      stderr=subprocess.PIPE)
+  (stdout, stderr) = process.communicate(input=input)
+  exit_code = process.wait()
+  if exit_code != 0 and error_message is not None:
+    if error_message is not None:
+      sys.stderr.write("bazelisk: {}\n".format(error_message))
+    write_binary_to_stderr(stderr)
+  return SubprocessResult(exit_code=exit_code)
+
+
+def write_binary_to_stderr(bytestring):
+  # Python 2 compatibility hack. In Python 3, you can't write byte
+  # strings to stdio; instead, you have to use the `sys.stderr.buffer`
+  # attribute, which is not available in Python 2.
+  buffer = getattr(sys.stderr, "buffer", sys.stderr)
+  buffer.write(bytestring)
+
+
+def verify_authenticity(binary_path, signature_path):
+  """Authenticate a binary and signature against the Bazel public key.
+
+  This will use a fresh temporary keyring populated only with the
+  Bazel team's signing key; it is independent of any existing PGP data
+  or settings that the user may have.
+
+  Args:
+    binary_path: File path to the Bazel binary to be executed.
+    signature_path: File path to the detached signature made by the
+      Bazel release PGP key to sign the provided binary.
+
+  Returns:
+    True if the binary is valid or gpg is not installed; False if gpg is
+    installed but we cannot determine that the binary is valid.
+  """
+  if subprocess_run(
+      ["gpg", "--batch", "--version"],
+      error_message=
+      "Warning: skipping authenticity check because GPG is not installed.",
+  ).exit_code != 0:
+    return True
```

Comment:
I think there should be a "project-owner" way of turning this into an error by default in order to use this on CI systems, for example, with confidence that the build is always built with an untainted bazel. I think using the `.bazelversion` file  to specify a flag, something like `--require-gpg-verification`, could do the trick nicely.